python/ray/data/_internal/execution/operators/limit_operator.py

```diff
@@ -0,0 +1,92 @@
+import ray
+import copy
+from typing import (
+    Deque,
+    List,
+    Optional,
+)
+from collections import deque
+from ray.data.block import (
+    Block,
+    BlockAccessor,
+    BlockMetadata,
+)
+from ray.data._internal.stats import StatsDict
+from ray.data._internal.execution.interfaces import (
+    PhysicalOperator,
+    RefBundle,
+)
+from ray.data._internal.remote_fn import cached_remote_fn
+from ray.types import ObjectRef
+
+
+class LimitOperator(PhysicalOperator):
+    """Physical operator for limit."""
+
+    def __init__(
+        self,
+        limit: int,
+        input_op: PhysicalOperator,
+    ):
+        self._limit = limit
+        self._consumed_rows = 0
+        self._buffer: Deque[RefBundle] = deque()
+        self._name = f"LimitOperator[limit={limit}]"
+        self._output_metadata: List[BlockMetadata] = []
+        self._num_outputs_total = input_op.num_outputs_total()
+        if self._num_outputs_total is not None:
+            self._num_outputs_total = min(self._num_outputs_total, limit)
+        super().__init__(self._name, [input_op])
+
+    def _limit_reached(self) -> bool:
+        return self._consumed_rows >= self._limit
+
+    def add_input(self, refs: RefBundle, input_index: int) -> None:
+        assert not self.completed()
+        assert input_index == 0, input_index
+        if self._limit_reached():
+            return
+        out_blocks: List[ObjectRef[Block]] = []
+        out_metadata: List[BlockMetadata] = []
+        for block, metadata in refs.blocks:
+            num_rows = metadata.num_rows
+            assert num_rows is not None
+            if self._consumed_rows + num_rows <= self._limit:
+                self._consumed_rows += num_rows
+                out_blocks.append(block)
+                out_metadata.append(metadata)
+                self._output_metadata.append(metadata)
+            else:
+                # Slice the last block.
+                num_rows_to_take = self._limit - self._consumed_rows
+                self._consumed_rows = self._limit
+                block = BlockAccessor.for_block(ray.get(block)).slice(
```

Comment:
We should still run it in a remote task to avoid needing to fetch the data block locally. This avoids the ray.put() later on.

Reply:
the purpose of doing this is to avoid putting too much data on the driver side? Or is there any other reason why we shouldn't do ray.put?

Reply:
Yeah, avoiding fetching large blocks to the driver is important, and it also avoids an extra data copy. If the data is split in a remote task, then only one put() happens, instead of two.